### PR TITLE
[GPU] Improve MobileNet v2 performance on Mali G77

### PR DESCRIPTION
**Pull request by antiagainst**
This commit tweaks (depthwise) convolution tiling size
and workgroup sizes to make MobileNet v2 performance
better on Mali G77. Along the way, we need to include
a few canonicalization patterns here and there to
enable proper folding and forwarding.

With this patch, MobileNet v2 fp32 takes ~20ms for
one inference, compared to ~90ms previously.

Command to generate the VM FlatBuffer

```
bazel run //integrations/tensorflow/e2e/keras/applications:applications_test_manual -- \
  --target_backends=tf,iree_vulkan,tflite \
  --data=imagenet \
  --model=MobileNetV2

iree-translate \
  -iree-mlir-to-vm-bytecode-module \
  -iree-hal-target-backends=vulkan-spirv \
  -iree-vulkan-target-triple="valhall-g77-unknown-android10" \
  -iree-spirv-enable-vectorization
  /tmp/iree/modules/MobileNetV2/iree_input.mlir
```

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp [hunk 1]
```diff
@@ -360,6 +364,38 @@ class ProcessPlaceHolder final
   }
 };
 
+// Folds linalg.reshape op that directly reshaping an iree.placeholder op into
+// the iree.placeholder op itself.
+class FoldReshapeIntoPlaceholder final
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:369]: It seems odd that this is here. Reshapes should be folded with the `hal.interface.*` operations after/during HLOToLinalgOnTensors or after HLOToLinalgOnBuffers. Is there an extra reshape inserted later than that?

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp [hunk 2]
```diff
@@ -360,6 +364,38 @@ class ProcessPlaceHolder final
...
+    if (!placeholderOp) return failure();
+    rewriter.replaceOpWithNewOp<IREE::PlaceholderOp>(
+        reshapeOp, reshapeOp.getResultType(), ValueRange(),
+        placeholderOp.getAttrs());
+    return success();
+  }
+};
+
+/// Removes iree.placeholder ops that has no uses.
+class RemoveDeadPlaceholder final
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:385]: This pass should do this automatically https://github.com/google/iree/blob/main/iree/compiler/Conversion/Common/RemoveDeadMemAllocsPass.cpp

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice. Few minor comments, but over looks great!

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:369]: This is generated when we convert depthwise mhlo.conv into linalg.depthwise_conv op in HLOToLinalgOnBuffers pass. I don't see we have a pattern there to fold linalg.reshape into iree.placeholder?

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:369]: Just add this pattern there?

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:385]: That's a pass though. Here the unused iree.placeholder is left after running by the above reshape folding pattern. So I need a pattern to clean it up before running next steps. So cannot rely on that pass. I can create a `populateRemoveDeadMemAllocsPatterns` entry point to pull that pattern, but not sure it's worth it given this pattern is just 4 LOC canonicalization.

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp:385]: If you move the above pattern to HLOToLinalgOnBuffers, and run this pass before TileAndFuse pass that should work

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp]: SG. Moved.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks!

**Pull request merged**